cmd/snap-confine/udev-support.c

```diff
@@ -18,290 +18,368 @@
 
 #include <ctype.h>
 #include <errno.h>
-#include <limits.h>
-#include <sys/sysmacros.h>
-#include <sched.h>
+#include <fcntl.h>
 #include <string.h>
 #include <sys/stat.h>
+#include <sys/sysmacros.h>
 #include <sys/types.h>
-#include <sys/wait.h>
 #include <unistd.h>
 
+#include <libudev.h>
+
+#include "../libsnap-confine-private/cleanup-funcs.h"
 #include "../libsnap-confine-private/snap.h"
 #include "../libsnap-confine-private/string-utils.h"
+#include "../libsnap-confine-private/cgroup-support.h"
 #include "../libsnap-confine-private/utils.h"
 #include "udev-support.h"
 
-static void
-_run_snappy_app_dev_add_majmin(struct snappy_udev *udev_s,
-			       const char *path, unsigned major, unsigned minor)
+__attribute__((format(printf, 2, 3)))
+static void sc_dprintf(int fd, const char *format, ...);
+
+static void sc_dprintf(int fd, const char *format, ...)
 {
-	int status = 0;
-	pid_t pid = fork();
-	if (pid < 0) {
-		die("cannot fork support process for device cgroup assignment");
-	}
-	if (pid == 0) {
-		uid_t real_uid, effective_uid, saved_uid;
-		if (getresuid(&real_uid, &effective_uid, &saved_uid) != 0)
-			die("cannot get real, effective and saved user IDs");
-		// can't update the cgroup unless the real_uid is 0, euid as
-		// 0 is not enough
-		if (real_uid != 0 && effective_uid == 0)
-			if (setuid(0) != 0)
-				die("cannot set user ID to zero");
-		char buf[64] = { 0 };
-		// pass snappy-add-dev an empty environment so the
-		// user-controlled environment can't be used to subvert
-		// snappy-add-dev
-		char *env[] = { NULL };
-		if (minor == UINT_MAX) {
-			sc_must_snprintf(buf, sizeof(buf), "%u:*", major);
-		} else {
-			sc_must_snprintf(buf, sizeof(buf), "%u:%u", major,
-					 minor);
-		}
-		debug("running snap-device-helper add %s %s %s",
-		      udev_s->tagname, path, buf);
-		// This code runs inside the core snap. We have two paths
-		// for the udev helper.
-		//
-		// First try new "snap-device-helper" path first but
-		// when running against an older core snap fallback to
-		// the old name.
-		if (access("/usr/lib/snapd/snap-device-helper", X_OK) == 0)
-			execle("/usr/lib/snapd/snap-device-helper",
-			       "/usr/lib/snapd/snap-device-helper", "add",
-			       udev_s->tagname, path, buf, NULL, env);
-		else if (access("/usr/libexec/snapd/snap-device-helper", X_OK) == 0)
-			execle("/usr/libexec/snapd/snap-device-helper",
-			       "/usr/libexec/snapd/snap-device-helper", "add",
-			       udev_s->tagname, path, buf, NULL, env);
-		else
-			execle("/lib/udev/snappy-app-dev",
-			       "/lib/udev/snappy-app-dev", "add",
-			       udev_s->tagname, path, buf, NULL, env);
-		die("execl failed");
+	va_list ap;
+	va_start(ap, format);
+	if (vdprintf(fd, format, ap) == -1) {
```

Comment:
A scenario that I'm not sure can happen in practice, but what if we have short write?

Say, we write `c 5:2 rwm`, and it's cut short and we effectively write `c 5:2 rw`, then there's fewer permissions than we wanted, but since the setup denies everything upfront, and only `rw` gets enabled, it should be fine. Another scenario, if we write `c 5:`, which is incorrect, the kernel will return an error, so the operation fails. I suppose we could add a note here to describe the scenario that a short write which produced a valid string could go undetected, but the permission set would limited to those flags that were written.

OTOH, it's probably unlikely to get a short write.

Reply:
I don't think it can happen in practice when writing to this kernel interface. The kernel scans the whole request, which must be a complete request line and must parse entirely.

Reply:
Note that snap-device-helper will write in a trailing newline for each item written, but this does not and none of the callers add `\n`. The upstream documentation in https://www.kernel.org/doc/Documentation/cgroup-v1/devices.txt uses `echo` in its example (like snap-device-helper does). This is probably ok (it certainly is on my 5.4 kernel), but it is a change over the current behavior. I think it would be much easier to add the newline in our code rather then verifying omitting it is safe over all kernels snapd might run on.

To @bboozzoo's point, it is never wrong to verify things are correct and I would prefer we check for a short write. Please note that snap-device-helper has `set -e`. OTOH I'm not sure what dash will do with a short write via `echo "$acl" > "$app_dev_cgroup/devices.allow"` when `set -e` is in use, but it sure seems plausible it would exit with error, and the old `_run_snappy_app_dev_add_majmin()` would `die()` in this case.

Lastly, honestly, I think that since sc_dprintf() is local to this file, why not simplify it to just be:
```
// sc_dprintf() writes the string (adding a trailing newline) to a file descriptor
static void sc_dprintf(int fd, const char *s)
{
	int n = dprintf(fd, "%s\n", s);
	if (n < 0 || n < strlen(s) + 1) {
		die("cannot write to fd %d", fd);
	}
}
```

If at some future date you want it to conditionally add the newline, just add a third arg.

Reply:
Sure, I'll handle both.

Reply:
I've hardened `sc_dprintf` but I dropped the strlen check. I think it's just the wrong way to implement this. Think about `%s` and empty string argument. Or about `%u` called with zero.

A while ago I implemented a full set of APIs that allow us to printf to open file descriptors, with nicer error checking but I never proposed that as it was a bit too NIH for snap-confine. I think the check we have now is enough.

Reply:
I had an idea how to harden this function without involving the format string length. Have a look please. :-)